flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamOperatorWrapper.java

```diff
@@ -0,0 +1,217 @@
+	 * {@link MailboxExecutor#yield()} to take the mails of closing operator and running timers and
+	 * run them.
+	 */
+	public void close(StreamTaskActionExecutor actionExecutor) throws Exception {
+		close(actionExecutor, false);
+	}
+
+	/**
+	 * Ends an input of the operator contained by this wrapper.
+	 *
+	 * @param inputId the input ID starts from 1 which indicates the first input.
+	 */
+	public void endOperatorInput(int inputId) throws Exception {
+		if (wrapped instanceof BoundedOneInput) {
+			((BoundedOneInput) wrapped).endInput();
+		} else if (wrapped instanceof BoundedMultiInput) {
+			((BoundedMultiInput) wrapped).endInput(inputId);
+		}
+	}
+
+	public OP getStreamOperator() {
+		return wrapped;
+	}
+
+	void setPrevious(StreamOperatorWrapper previous) {
+		this.previous = previous;
+	}
+
+	void setNext(StreamOperatorWrapper next) {
+		this.next = next;
+	}
+
+	private void close(StreamTaskActionExecutor actionExecutor, boolean invokingEndInput) throws Exception {
+		if (invokingEndInput) {
+			// NOTE: This only do for the case where the operator is one-input operator. At present,
+			// any non-head operator on the operator chain is one-input operator.
+			actionExecutor.runThrowing(() -> endOperatorInput(1));
+		}
+
+		closeOperatorAndQuiesceTimeService(actionExecutor);
+
+		// propagate the close operation to the next wrapper
+		if (next != null) {
+			next.close(actionExecutor, true);
+		}
+	}
+
+	private void closeOperatorAndQuiesceTimeService(StreamTaskActionExecutor actionExecutor)
+		throws InterruptedException, ExecutionException {
+
+		final CompletableFuture<Void> closedFuture = new CompletableFuture<>();
+
+		// 1. executing the close operation must be deferred to the mailbox to ensure that mails already
+		//    in the mailbox are finished before closing the operator
+		// 2. to ensure that there is no longer output triggered by the timers before invoke the "endInput"
+		//    methods of downstream operators in the operator chain, we must quiesce the processing time
+		//    service to prevent the pending timers from firing, but allow the timers in running to finish
+		// 3. when the second step is finished, send a closed mail to ensure that the timers yet in mailbox
+		//    are finished before exiting the following mailbox processing loop
+		// TODO: To ensure the strict semantics of "close", the operator should be allowed to decide how to
+		//  handle (cancel or trigger) the pending timers before being closed, and the second step should be
+		//  finished before the first step
+		deferCloseOperatorToMailbox(actionExecutor)
+			.thenCompose(unused  -> quiesceProcessingTimeService())
+			.thenRun(() -> sendClosedMail(closedFuture))
+			.exceptionally((throwable) -> forwardException(throwable, closedFuture));
+
+		// run the mailbox processing loop until the closing operation is finished
+		while (!closedFuture.isDone()) {
+			mailboxExecutor.yield();
```

Comment:
Regarding deadlock if methods are chained without passing future as a parameter:
`thenCompose` should have been used in both places (I previously suggested it only for `sendClosedMail`, sorry for misleading).

Reply:
Using the following methods can also cause deadlocks.
```
	private void closeOperatorAndQuiesceTimeService(StreamTaskActionExecutor actionExecutor)
		throws InterruptedException, ExecutionException {

		CompletableFuture<Void> closedFuture = deferCloseOperatorToMailbox(actionExecutor)
			.thenCompose(unused  -> quiesceProcessingTimeService())
			.thenCompose(unused -> sendClosedMail());

		// run the mailbox processing loop until the closing operation is finished
		while (!closedFuture.isDone()) {
			mailboxExecutor.yield();
		}
                ......
	}

	private CompletableFuture<Void> sendClosedMail() {
		final CompletableFuture<Void> future = new CompletableFuture<>();

		mailboxExecutor.execute(() ->
			future.complete(null), "StreamOperatorWrapper#sendClosedMail for " + wrapped);
		return future;
	}
```
When the execution of `sendClosedMail()`  in `yield() ` is completed and returns to `while (!closedFuture.isDone()) ` ,  `closedFuture` may not be set to  the `done` state（`thenCompose()` is different from `theRun()` at this point）, which causes `yield()` to be called again, and at this time, all mails of the operator have been processed.

Reply:
Could you please provide the test which fails?

As I understand from `CompletableFuture`, dependent completions in our case are all run by the same task thread - immediately after it completes the 1st future.
So returning from `yield` should mean that the future is done.

If there **IS** a chance of deadlock, I think we should guard against it, e.g. by adding a dummy mail before yielding.

Reply:
> Could you please provide the test which fails?

In the local computer, by running `StreamOperatorWrapperTest` 100 times repeatedly, a deadlock occurs when `StreamOperatorWrapperTest#testClose` is executed at a certain time.

> If there IS a chance of deadlock, I think we should guard against it, e.g. by adding a dummy mail before yielding.

Using `theRun()` does not result in deadlock. In addition, the closed mail is also a dummy mail.

Reply:
Ok, I see: one of the futures in the chain can be completed by the timer thread, and the resulting future can also be completed by it, i.e. after `yield`. 

In this case, we can just alternate between trying to `yield` and checking the future:
```
final CompletableFuture<Void> closedFuture =
    deferCloseOperatorToMailbox(actionExecutor)
        .thenCompose(unused -> quiesceProcessingTimeService())
        .thenCompose(unused -> sendClosedMail());

while (mailboxExecutor.tryYield() || !closedFuture.isDone()) {} 
```
(I was able to reproduce the deadlock and to eliminate it with the above approach)

Reply:
This approach can also eliminate deadlock, but there will be some NOOP loops that occupy unnecessary CPU.

Reply:
This is not an issue I think since the job is shutting down.

Reply:
But this can affect other jobs running on the same  machine.